larq/optimizers_v2.py

```diff
@@ -1,10 +1,121 @@
+import warnings
+
 import tensorflow as tf
 import larq as lq
 
 from larq import utils
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """A case operation for optimizers that each train a subset of a model's variables.
+
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer.
+
+    # Arguments
+    pred_opt_pairs: a list of `(tf.keras.optimizers.Optimzer, pred)` pairs, where `pred` 
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) === True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
+    """
+
+    def __init__(self, pred_opt_pairs, default, name="optim_case"):
+        super().__init__(name=name)
+
+        for i, (pred, opt) in enumerate(pred_opt_pairs):
+            if not callable(pred):
+                raise TypeError(
+                    f"Expected callable predicate at `pred_opt_pairs[{i}][0]` but got `{type(pred)}`."
+                )
+            if not isinstance(opt, tf.keras.optimizers.Optimizer):
+                raise TypeError(
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(opt)}`."
+                )
+        self.pred_opt_pairs = pred_opt_pairs
+
+        if default is not None and not isinstance(
+            default, tf.keras.optimizers.Optimizer
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+        self.default = default
+
+    def __getattr__(self, name):
+        if name == "lr":
+            raise ValueError("No single learning rate available for CaseOptimizer.")
+        return super().__getattr__(name)
```

Comment:
Actually I don't think this will do anything since the optimizer won't have any interesting to attributes. I think it is OK to remove the whole `__getattr__` method. Or am I missing something?

Reply:
I mostly added it because Bop had it and I assumed it was getting called externally somewhere. If that's not the case I'm fine with removing it. Thoughts @koenhelwegen?